File: 0056-merge-intervals/0056-merge-intervals.py
class Solution:
    def merge(self, intervals: List[List[int]]) -> List[List[int]]:
        result = []
        intervals.sort(key=lambda x: x[0])
        n = len(intervals)
        for interval in intervals:
            if not result or (result[-1][1] < interval[0] or interval[1] < result[-1][0]):
                result.append(interval)
            else:
                result[-1][0] = min(result[-1][0], interval[0])
                result[-1][1] = max(result[-1][1], interval[1])
        return result
            

File: 0056-merge-intervals/0056-merge-intervals.java
class Solution {
    public int[][] merge(int[][] intervals) {
        List<int[]> result = new ArrayList<>();

        Arrays.sort(intervals, (i1, i2) -> Integer.compare(i1[0], i2[0]));

        for(int[] interval: intervals){
            if(result.isEmpty() || result.get(result.size()-1)[1] < interval[0]){
                result.add(interval);
            }else{
                //merge the intervals
                int[] oldInterval = result.get(result.size()-1);
                oldInterval[0] = Math.min(oldInterval[0], interval[0]);
                oldInterval[1] = Math.max(oldInterval[1], interval[1]);
                // System.out.println(oldInterval[0] + " " + oldInterval[1]);
                result.set(result.size()-1, oldInterval);
            }
        }
        return result.toArray(new int[result.size()][]);
    }
}